org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/handler/ZigBeeSerialPort.java

```diff
@@ -279,15 +278,45 @@ public void serialEvent(SerialPortEvent event) {
         if (event.getEventType() == SerialPortEvent.DATA_AVAILABLE) {
             try {
                 synchronized (bufferSynchronisationObject) {
-                    int recv;
-                    while ((recv = inputStream.read()) != -1) {
-                        buffer[end++] = recv;
+                    int available = inputStream.available();
+                    logger.trace("Processing DATA_AVAILABLE event: have {} bytes available", available);
+                    byte[] buf = new byte[available];
+                    int offset = 0;
+                    while (offset != available) {
+                        if (logger.isTraceEnabled()) {
+                            logger.trace("Processing DATA_AVAILABLE event: try read  {} at offset {}",
+                                    available - offset, offset);
+                        }
+                        int n = inputStream.read(buf, offset, available - offset);
+                        if (logger.isTraceEnabled()) {
+                            logger.trace("Processing DATA_AVAILABLE event: did read {} of {} at offset {}", n,
+                                    available - offset, offset);
+                        }
+                        if (n <= 0) {
+                            logger.error("Processing DATA_AVAILABLE event: Expected to be able to read " + available
```

Comment:
Please can we drop the logging level down to debug (or possibly warn). Also, please use logging with individual parameters rather than string concatenation.

Reply:
In fact I'd probably suggest to remove this log entry since there is an exception that also provides the same information.

Reply:
This situation ought never to occur; the amount being read has been determined to be available, so if it can't be read it probably means that another thread or process is simultaneously  reading from the port, which is likely more serious than a warning.

Reply:
Within the OH context though, this is only meant for issues that are system wide. I would really prefer to reduce this level, or simply remove it since there is already an exception thrown which is caught and logged so there seems little point in logging the same thing twice does there?

Reply:
I don't know what other  IOExceptions, if any, might have previously been   masked here. 

But I will  stop logging this error as an error, and make the logging statement in the catch clause log at warning. 

```
/var/log/openhab2/openhab.log.5:2019-10-18 14:55:34.553 [ERROR] [ding.zigbee.handler.ZigBeeSerialPort] - Serial Error: Port /dev/ttyUSB1 in use.
```

Reply:
Thanks, but we are still logging this statement twice - it's exactly the same string that is logged, then thrown as an exception and caught a few lines further down and then logged again. I don't think there is there really a need for this duplication?

Reply:
The change removed separately logging the error as an warning, and now logs all IOExceptions as warnings.